### PR TITLE
dns: provide generic dns_query() function

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Since we can have multiple DNS backends now, provide a common `dns_query()` wrapper function that allows to write backend agnostic applications.

### CODE REVIEW

File: sys/include/net/dns.h [hunk 1]
```diff
@@ -43,6 +51,41 @@ extern "C" {
...
+ * @param[out]  addr_out        buffer to write result into
+ * @param[in]   family          Either AF_INET, AF_INET6 or AF_UNSPEC
+ *
+ * @return      the size of the resolved address on success
+ * @return      < 0 otherwise
+ */
+static inline int dns_query(const char *domain_name, void *addr_out, int family)
+{
+#if defined(MODULE_SOCK_DNS) || defined(MODULE_SOCK_DNS_MOCK)
+    return sock_dns_query(domain_name, addr_out, family);
```
Comment (miri64) [inline sys/include/net/dns.h:79]: Why give `sock_dns` precedence over the encrypted versions, if both are compiled in?

File: sys/include/net/dns.h [hunk 2]
```diff
@@ -43,6 +51,41 @@ extern "C" {
...
+#elif defined(MODULE_SOCK_DODTLS)
+    return sock_dodtls_query(domain_name, addr_out, family);
+#elif defined(MODULE_GCOAP_DNS)
+    return gcoap_dns_query(domain_name, addr_out, family);
+#else
+    return -ENOTSUP;
+#endif
```
Comment (miri64) [inline sys/include/net/dns.h:86]: Also: what if multiple DNS over X variants are compiled in, but some have a server configured and some have not?

Comment (benpicco) [reply sys/include/net/dns.h]: Good point, should be fixed now.

File: sys/include/net/dns.h (hunk 1 above)
Comment (benpicco) [reply sys/include/net/dns.h]: Re-shuffled them

File: sys/include/net/dns.h [hunk 3]
```diff
@@ -20,6 +20,16 @@
 #ifndef NET_DNS_H
 #define NET_DNS_H
 
+#if defined(MODULE_SOCK_DNS) || defined(MODULE_SOCK_DNS_MOCK)
```
Comment (miri64) [inline sys/include/net/dns.h:23]:
```suggestion
#include "kernel_defines.h"

```

File: sys/include/net/dns.h [hunk 4]
```diff
@@ -20,6 +20,16 @@
...
+#include "net/sock/dns.h"
+#endif
+#if defined(MODULE_SOCK_DODTLS)
```
Comment (miri64) [inline sys/include/net/dns.h:26]:
```suggestion
```

File: sys/include/net/dns.h [hunk 5]
```diff
@@ -20,6 +20,16 @@
...
+#include "net/sock/dodtls.h"
+#endif
+#if defined(MODULE_GCOAP_DNS)
```
Comment (miri64) [inline sys/include/net/dns.h:29]:
```suggestion
```

File: sys/include/net/dns.h [hunk 6]
```diff
@@ -20,6 +20,16 @@
...
+#include "net/gcoap/dns.h"
+#endif
```
Comment (miri64) [inline sys/include/net/dns.h:31]:
```suggestion
```

File: sys/include/net/dns.h [hunk 7]
```diff
@@ -43,6 +53,51 @@ extern "C" {
...
+ */
+static inline int dns_query(const char *domain_name, void *addr_out, int family)
+{
+    int res = -ENOTSUP;
+
+#if defined(MODULE_GCOAP_DNS)
+    if (res < 0) {
+        res = gcoap_dns_query(domain_name, addr_out, family);
+    }
+#endif
```
Comment (miri64) [inline sys/include/net/dns.h:86]:
```suggestion
    if (IS_USED(MODULE_GCOAP_DNS) && (res < 0)) {
        res = gcoap_dns_query(domain_name, addr_out, family);
    }
```

File: sys/include/net/dns.h [hunk 8]
```diff
@@ -43,6 +53,51 @@ extern "C" {
...
+#if defined(MODULE_SOCK_DODTLS)
+    if (res < 0) {
+        res = sock_dodtls_query(domain_name, addr_out, family);
+    }
+#endif
```
Comment (miri64) [inline sys/include/net/dns.h:91]:
```suggestion
    if (IS_USED(MODULE_SOCK_DODTLS) && res < 0) {
        res = sock_dodtls_query(domain_name, addr_out, family);
    }
```

File: sys/include/net/dns.h [hunk 9]
```diff
@@ -43,6 +53,51 @@ extern "C" {
...
+#if defined(MODULE_SOCK_DNS) || defined(MODULE_SOCK_DNS_MOCK)
+    if (res < 0) {
+        res = sock_dns_query(domain_name, addr_out, family);
+    }
+#endif
```
Comment (miri64) [inline sys/include/net/dns.h:96]:
```suggestion
    if ((IS_USED(MODULE_SOCK_DNS) || IS_USED(MODULE_SOCK_DNS_MOCK)) && res < 0) {
        res = sock_dns_query(domain_name, addr_out, family);
    }
```

**Review by miri64 — COMMENTED** [review]
`sock_dtls` is a bit fiddly when it comes to this (should be fixed long-term!), but does this work? If yes it makes the code a bit more easy to read. If does not work, `IS_USED` is preferred in any case over `defined` to check the usage of a module.

File: sys/include/net/dns.h [hunk 10]
```diff
@@ -43,6 +53,51 @@ extern "C" {
...
+ *
+ * @return      the size of the resolved address on success
+ * @return      < 0 otherwise
+ */
+static inline int dns_query(const char *domain_name, void *addr_out, int family)
+{
+    int res = -ENOTSUP;
+
+#if defined(MODULE_GCOAP_DNS)
+    if (res < 0) {
```
Comment (miri64) [inline sys/include/net/dns.h:83]: IMHO it could make sense to use `res <= 0` here.

Comment (miri64) [reply sys/include/net/dns.h:83]: BTW, we should rethink this, once we expand the API e.g. for DNS-SD or other DNS records than A/AAAA in general.

File: sys/include/net/dns.h (hunk 9 above)
Comment (benpicco) [reply sys/include/net/dns.h:96]: I really don’t see what’s the advantage of `IS_USED` outside C code

**Comment by benpicco** [comment]
Including `dodtls.h` unconditionally fails:

```
sys/include/net/sock/dtls.h:1001:10: fatal error: sock_dtls_types.h: No such file or directory
 1001 | #include "sock_dtls_types.h"
```

> If does not work, `IS_USED` is preferred in any case over defined to check the usage of a module.

But why? I see no advantage of using `IS_USED` outside C code, in fact it makes the code less idiomatic.

> benpicco deleted a comment from miri64 18 minutes ago

Huh I think I did delete a comment of mine 😅

**Comment by miri64** [comment]
> > 
> 
> But why? I see no advantage of using `IS_USED` outside C code, in fact it makes the code less idiomatic.

AFAIK, `defined` merely checks if the macro is defined, `IS_USED` also checks if the macro is either 0 or 1.

**Comment by benpicco** [comment]
Yes but under which conditions would be have `MODULE_GCOAP_DNS` defined to 0?

**Comment by miri64** [comment]
> Yes but under which conditions would be have `MODULE_GCOAP_DNS` defined to 0?

Maybe testing, where you want to have that module in, but you don't want the behavior it implies in other modules, so you set `CFLAGS += -DMODULE_GCOAP_DNS=0`... but granted this very pretty constructed and might lead to other conflicts... so idk.

**Comment by miri64** [comment]
> > Yes but under which conditions would be have `MODULE_GCOAP_DNS` defined to 0?
> 
> Maybe testing, where you want to have that module in, but you don't want the behavior it implies in other modules, so you set `CFLAGS += -DMODULE_GCOAP_DNS=0`... but granted this very pretty constructed and might lead to other conflicts... so idk.

But to wrap up this discussion: using C-conditionals is not possible, because `sock/dtls.h` (which is included by `gcoap_dtls` and `sock_dodtls` headers) requires a `sock_dtls_types.h` to be present, which is only provided if a `sock_dtls` implementation is present. As such, C-conditionals can't be used at the moment (will try if I can do a quick fix). This only leaves https://github.com/RIOT-OS/RIOT/pull/18378#pullrequestreview-1053016185.

**Comment by miri64** [comment]
> (will try if I can do a quick fix)

See https://github.com/RIOT-OS/RIOT/pull/18380. @benpicco can you try, if with that C-conditionals work?

**Comment by benpicco** [comment]
Yup that works

**Comment by miri64** [comment]
> Yup that works

Than I would be very happy about an ACK in #18380 ;-)

**Comment by benpicco** [comment]
It's a two line change, can't I just include it here?

**Comment by miri64** [comment]
As long as authorship is preserved.. But even for 2 line change: I won't review my own code.

**Comment by benpicco** [comment]
> But even for 2 line change: I won't review my own code.

Are you being serious? :smile:

**Comment by miri64** [comment]
> > But even for 2 line change: I won't review my own code.
> 
> Are you being serious? :smile:
> And anyway, if I cherry pick the commit, isn't that pretty much me giving it an ACK?

Yes and yes. If I f*** up my own code and the reviewer did not test properly, the responsibility does not only lie with me. If I f*** my own code and ACK my own code with only the testing I did, the responsibility is completely with me. So, knowing myself, I'd rather  avoid the risk of that :wink:.

**Comment by benpicco** [comment]
Now this PR only contains commits by yours truly, you are free to review it :wink:

File: sys/net/netutils/util.c [hunk 11]
```diff
@@ -20,9 +20,9 @@
 #include <string.h>
 
 #include "net/utils.h"
-#if defined(MODULE_SOCK_DNS) || defined(MODULE_SOCK_DNS_MOCK)
+#if defined(MODULE_DNS)
```
Comment (miri64) [inline sys/net/netutils/util.c:23]: Why is this `#if` necessary at all?

File: sys/net/sock/sock_util.c [hunk 12]
```diff
@@ -29,8 +29,8 @@
 #include "net/sock/udp.h"
 #include "net/sock/util.h"
 
-#if defined(MODULE_SOCK_DNS) || defined(MODULE_SOCK_DNS_MOCK)
-#include "net/sock/dns.h"
+#if defined(MODULE_DNS)
```
Comment (miri64) [inline sys/net/sock/sock_util.c:32]: Same here.

File: sys/Makefile.dep [hunk 13]
```diff
@@ -602,6 +602,10 @@ ifneq (,$(filter sock_dns,$(USEMODULE)))
   USEMODULE += posix_headers
 endif
 
+ifneq (,$(filter sock_dns_mock,$(USEMODULE)))
+  USEMODULE += dns
+endif
```
Comment (miri64) [inline sys/Makefile.dep:607]: If I am not mistaken, this dependency might become unnecessary with the `#if defined(MODULE_DNS)`s removed

Comment (benpicco) [reply sys/Makefile.dep:607]: The problem is that compilation then fails if the `sock` module is not used

```
In file included from /home/benpicco/dev/RIOT/sys/include/net/sock/dns.h:33,
                 from /home/benpicco/dev/RIOT/sys/include/net/dns.h:24,
                 from /home/benpicco/dev/RIOT/sys/net/netutils/util.c:24:
/home/benpicco/dev/RIOT/sys/include/net/sock/udp.h:788:10: fatal error: sock_types.h: No such file or directory
  788 | #include "sock_types.h"
```

Comment (miri64) [reply sys/Makefile.dep:607]: Mh, so similar problem to `sock_dtls`... but let's not stall this any further and merge as is.

**Review by miri64 — APPROVED** [review]
ACK

**Pull request merged**